feat: US131353 - add a11y features to count-badge (#1646)
BrightspaceUI/core: components/count-badge/README.md

| `type`, default: `count` | String | The type of the badge. Valid options are `"notification"` and `"count"`. Notification badges are orange and count badges are grey. |
| `max-digits`, default: `2` when `type="notification"` | Number | Optionally specify a digit limit, after which numbers are truncated. Defaults to two for `"notification"` type and no limit for `"count"` type.
| `hide-zero`, default: `false` | Boolean | Optionally choose not to show the count badge when the number is zero. |
| `text`, required | String | Descriptive text for the badge which will act as an accessible label and tooltip text when tooltips are enabled. |

> | `tab-stop`, default: `false` | Boolean | Optionally choose to make the badge a tab stop. |

Reviewer: Just wrapping my head around the need for both `tab-stop` and `has-tooltip`... sometimes we want to be able to tab to a count-badge so the accessibility label can be read, but we _don't_ want a tooltip?

I'm wondering if these two should be combined for a keyboard-only user - basically, if you want to add a tab-stop for screenreader users, then keyboard-only users also benefit from the tooltip.  And the reverse seems to already be true - if you add a tooltip, then you're already adding a tab-stop.

Author: The spec mentions we can do a tab stop without a tooltip. I think one case with a tab stop and no tooltip could be portfolio where the text is already showing something similar to what the tooltip would show (ie. overdue evidence). A screenreader may want to still read out "X pieces of evidence" but not show a tooltip there.

> | `announce-changes`, default: `false` | Boolean | Optionally choose to announce changes to the badge with an aria-live region. If the number property is changed, the text will be read by screen-readers. |

Reviewer: ```suggestion
| `announce-changes`, default: `false` | Boolean | Optionally choose to announce changes to the badge with an aria-live region. If the text property is changed, the text will be read by screen-readers. |
```
I think this is technically more accurate with the latest implementation

| `has-tooltip`, default: `false` | Boolean | Optionally choose to have a tooltip below the badge. |

### Accessibility Properties

To make your `d2l-count-badge` accessible, use the following properties when applicable:

| Attribute | Description |
|--|--|
| `text`, required | Only the text will be read by screen-readers (not the number), so include the number in the text. |

> | `tab-stop` | A tab stop allows keyboard users to easily tab to the badge. Otherwise, keyboard users will need to arrow through to the badge. |

Reviewer: ```suggestion
| `tab-stop` | A tab stop allows screen-reader users to easily tab to the badge. Otherwise, screen-reader users will need to arrow through to the badge. |
```

Wonder if this is more accurate?  (Since just the tab-stop doesn't really add anything for keyboard users)

| `announce-changes` | Use "announce-changes" if screen-reader users should be notified that the count has been updated, such as a new notification. The "text" property will be read as soon as the screen-reader is idle. |

> | `tooltip` | The tooltip will be visible on hover, and read out by screen-readers. |

Reviewer: ```suggestion
| `has-tooltip` | The tooltip will be visible on hover/tab stop, and read out by screen-readers. |

```

| `max-digits`, default: `2` when `type="notification"` | Number | Optionally specify a digit limit, after which numbers are truncated. Defaults to two for `"notification"` type and no limit for `"count"` type.
| `hide-zero`, default: `false` | Boolean | Optionally choose not to show the count badge when the number is zero. |

### Accessibility Properties

To make your `d2l-count-badge` accessible, use the following properties when applicable:

| Attribute | Description |
|--|--|

> | `description` | Optionally choose to add a description, to be the aria-label on the badge. |

Reviewer 2: Wondering: without this, a screen reader user is just going to happen upon a random number when combing through the page. Should we rename this to `label` and make it required?

Author: In its current state, without the tab stop, the label just never gets read. There are some cases, like card-footer-link, where screen readers don't read the count badge at all, so no label is required.

Reviewer 2: What about just arrowing through? It should get read then. Like most non-interactive things on the page (headings, paragraphs, lists, etc) don't get tab stops either -- the screen reader user navigates to them using other means.

Author: I see what you mean now. Then it does make sense to make it required, I'll send an update with the *hopefully* fixed tests.

> | `description`, required | A description of the badge for screen-readers, to be used as an aria-label. NOTE: Only the description will be read by screen-readers (not the number). |

Reviewer 2: To me, `description` still sounds a little... optional. Wondering if `text` might be more appropriate? That would line up nicely with `<d2l-button-icon>`, which is kind of a similar component.

| `tab-stop` | Optionally choose to make the badge a tab stop. |

> | `announce-changes` | Optionally choose to announce changes to the badge with an aria-live region. If the number property is changed, the description will be read by screenreaders. |

Reviewer 2: I think we typically still include accessibility-related properties in the property list above, and then just call them out again in these lists with extra guidance. Although I'm not sure if @margaree has thoughts here or if we've changed our approach with the Daylight site?

Reviewer 3: Yup that's still correct
